tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,455 @@
+      nodes_from_other_layers=nodes_from_other_layers)
+
+
+def _handle_nodes_from_other_layers(current_spec,
+                                    execution_specs,
+                                    graph,
+                                    node_name_to_node_def):
+  """Handle nodes that are from other layers."""
+  for node_name in current_spec.nodes_from_other_layers:
+    for previous_spec in execution_specs:
+
+      if node_name in previous_spec.body_node_names:
+        previous_spec.output_names.add(node_name)
+        current_spec.input_names.add(node_name)
+
+        node = node_name_to_node_def[node_name]
+        placeholder = _create_placeholder_node_from_existing_node(node, graph)
+        current_spec.subgraph.node.append(placeholder)
+
+  return current_spec
+
+
+def _partition_one_remote_op_layer(remote_op_names, node_name_to_input_names):
+  """Construct execution spec for remote ops."""
+  list_of_specs = []
+  for remote_op_name in remote_op_names:
+    spec = ExecutionSpec(
+        subgraph=None,
+        input_names=set(node_name_to_input_names[remote_op_name]),
+        output_names=set([remote_op_name]),
+        is_remote_op=True,
+        body_node_names=set([remote_op_name]),
+        nodes_from_other_layers=set([]))
+    list_of_specs.append(spec)
+
+  return list_of_specs
+
+
+def _create_placeholder_node(dtype, shape, name):
+  temp = tf.Graph()
+  with temp.as_default():
+    tf.compat.v1.placeholder(dtype=dtype, shape=shape, name=name)
+    return temp.as_graph_def().node[0]  # The first and the only node
+
+
+def _create_placeholder_node_from_existing_node(node, graph):
+  operation = graph.get_operation_by_name('import/%s' % (node.name))
+  dtype = operation.outputs[0].dtype
+  return _create_placeholder_node(dtype=dtype,
+                                  shape=None,
+                                  name=node.name)
+
+
+def _get_input_names_from_subgraph(subgraph):
+  input_names = {node.name for node in subgraph.node
+                 if _is_placeholder_op(node)}
+  return input_names
+
+
+def _get_body_node_names_from_subgraph(subgraph):
+  body_node_names = {node.name for node in subgraph.node
+                     if not _is_placeholder_op(node)}
+  return body_node_names
+
+
+class Relations:
+  """A class that outputs remote op layers (custom topological sort).
+
+  What is a layer? A layer is a set of remote ops that don't have
+    dependencies on each other and are ready to execute."""
+  def __init__(self, relations):
+    self.relations = relations
+    self.processed = set([])
+    self.to_be_processed = set(relations.keys())
+
+  def check_if_finished(self):
+    return not self.to_be_processed
+
+  def get_next_layer(self):
+    """Get the next set of remote ops."""
+    layer_nodes = set([])
+
+    for node in self.to_be_processed:
+      node_inputs = set(self.relations[node])
+      if node_inputs.issubset(self.processed):
```

Comment:
In this algorithm, I don't think you need you need to check _all the ancestors_ of the node. It suffices to check only the immediate parents. This is because of the following invariant: if a node is in the processed set, all of its ancestors must also be in the processed set.

This should allow you to simplify the algorithm.

Reply:
Indeed, the current algorithm looks for the immediate remote op parents. I changed some names and comments to clarify this.